config/config.go

```diff
@@ -166,14 +223,327 @@ func (cs configs) ResolveConfig(f func(configs []interface{}) error) error {
+		o.EC2RoleCredentialOptions = v
+		return nil
+	}
+}
+
+// getEndpointCredentialOptions returns the wrapped function to set endpointcreds.Options
+func (o LoadOptions) getEndpointCredentialOptions(context.Context) (func(*endpointcreds.Options), bool, error) {
+	if o.EndpointCredentialOptions == nil {
+		return func(*endpointcreds.Options) {}, false, nil
+	}
+
+	return o.EndpointCredentialOptions, true, nil
+}
+
+// WithEndpointCredentialOptions is a helper function to construct functional options
+// // that sets a function to use endpointcreds.Options on config's LoadOptions
+func WithEndpointCredentialOptions(v func(*endpointcreds.Options)) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.EndpointCredentialOptions = v
+		return nil
+	}
+}
+
+// getWebIdentityRoleCredentialOptions returns the wrapped function
+func (o LoadOptions) getWebIdentityRoleCredentialOptions(context.Context) (func(*stscreds.WebIdentityRoleOptions), bool, error) {
+	if o.WebIdentityRoleCredentialOptions == nil {
+		return func(*stscreds.WebIdentityRoleOptions) {}, false, nil
+	}
+
+	return o.WebIdentityRoleCredentialOptions, true, nil
+}
+
+// WithWebIdentityRoleCredentialOptions is a helper function to construct functional options
+// that sets a function to use stscreds.WebIdentityRoleOptions on config's LoadOptions
+func WithWebIdentityRoleCredentialOptions(v func(*stscreds.WebIdentityRoleOptions)) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.WebIdentityRoleCredentialOptions = v
+		return nil
+	}
+}
+
+// getAssumeRoleCredentialOptions returns AssumeRoleCredentialOptions from LoadOptions
+func (o LoadOptions) getAssumeRoleCredentialOptions(context.Context) (func(options *stscreds.AssumeRoleOptions), bool, error) {
+	if o.AssumeRoleCredentialOptions == nil {
+		return func(*stscreds.AssumeRoleOptions) {}, false, nil
+	}
+
+	return o.AssumeRoleCredentialOptions, true, nil
+}
+
+// WithAssumeRoleCredentialOptions  is a helper function to construct functional options
+// that sets a function to use stscreds.AssumeRoleOptions on config's LoadOptions
+func WithAssumeRoleCredentialOptions(v func(*stscreds.AssumeRoleOptions)) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.AssumeRoleCredentialOptions = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getHTTPClient(ctx context.Context) (HTTPClient, bool, error) {
+	if o.HTTPClient == nil {
+		return nil, false, nil
+	}
+
+	return o.HTTPClient, true, nil
+}
+
+// WithHTTPClient is a helper function to construct functional options
+// that sets HTTPClient on LoadOptions
+func WithHTTPClient(v HTTPClient) func(*LoadOptions) error {
+	return func(o *LoadOptions) error {
+		o.HTTPClient = v
+		return nil
+	}
+}
+
+func (o LoadOptions) getAPIOptions(ctx context.Context) ([]func(*middleware.Stack) error, bool, error) {
+	if o.APIOptions == nil {
+		return []func(stack *middleware.Stack) error{}, false, nil
+	}
+
+	return o.APIOptions, true, nil
+}
```

Comment:
should not initialize values for unset options.  Opens room for bugs because code didn't check the bool correctly.